File: fortran/euler002/euler002.f90
!    file:    euler002.f90
!    author:  Marvin Smith
!    date:    4/25/2015
!
!    Purpose:  Solves Project Euler 002 using Fortran 90
!

PROGRAM Euler002
    USE FibonacciModule
    IMPLICIT none
    INTEGER :: sum, counter, temp_value, max_fib_number, Fmax

!  Set running sum (Start with 2 as that is automatic)
sum = 2

!  Define the max value to find fibonacci numbers for
max_fib_number=4000000

!  Find the max Fn such that it is closest but less than max
Fmax = Find_Max_Fibonacci( max_fib_number )

!  Start Iterating over Fibonacci Values
do counter = 2, Fmax

    !  Call the fibonacci number for the current counter
    temp_value = Fibonacci(counter)
    
    !  Check if it is even
    if (MODULO(temp_value,2) == 0) then
        sum = sum + temp_value
    endif

end do

!  Print the Sum
print*, 'Sum:', sum

!  Exit Program
end

File: cpp/euler002/euler002.cpp
/**
 * @file    euler002.cpp
 * @author  Marvin Smith
 * @date    4/25/2015
*/
#include <iostream>

/**
 * Fibonacci Number
 */
int Fibonacci( int const& F )
{
    // Handle Exit Conditions
    if( F < 2 ){ return F; }

    // Otherwise
    return Fibonacci( F-1 ) + Fibonacci( F-2 );
}


/**
 * Find Max Fibonacci
 */
int Find_Max_Fibonacci( const int& max_value )
{
    // Start iteration
    int counter = 2;
    int Fmax = counter;
    while( true ){

        // Update FMax
        Fmax = counter-1;

        // Check counter
        if( Fibonacci(counter++) > max_value )
        {
            return Fmax;
        }

        // Error Check
        if( counter > 1000000 ){
            return -1;
        }
    }
    return -1;
}

/**
 * Main Function
 */
int main( int argc, char* argv[] )
{

    // Compute Max Euler Value for Given Input
    const int max_value = 4000000;
    const int Fmax = Find_Max_Fibonacci( max_value );

    // Compute Sum of Even Fibonacci Numbers
    int result = 0;
    int value;
    for( int i=0; i<=Fmax; i++ ){
        value = Fibonacci(i);
        if( value % 2 == 0 ){
            result += value;
        }
    }

    // Print Result
    std::cout << result << std::endl;

    return 0;
}


